### DISCUSSION TITLE
Is there a way to get time of modification and/or creation on files in micropython?

### DISCUSSION THREAD

**Discussion by fishjoe** [micropython/micropython#9698] [General]
Is there a way to get time of modification and/or creation on files in micropython?

many thanks.

**Comment by sosi-deadeye** [comment]
### tl;dr
To get the `atime`, `mtime` and `ctime` you need the last 3 elements of the result from `os.stat("path")`
A shortcut for this:
```python
import os

atime, mtime, ctime = os.stat("/boot.py")[-3:]
```
### timestamp > datetime-object
The existing [micropython-datetime](https://github.com/micropython/micropython-lib/tree/master/python-stdlib/datetime) library has no support for `fromtimestamp`, but a `datetimeobject` can deal with `addition` or `substraction` of `timedelta` objects.

On different platforms are different offsets used.
On an ESP32 it's: 2000-01-01 00:00 (I guessed it)

This can be used to calculate the offset, but first you need the library. You can install the library if your controller is connected to a wifi and has an internet connection:
```python
import mip
mip.install("datetime")
```

### Example Code
```python
import datetime
import os
import sys


if "ESP32" in sys.implementation._machine:
    OFFSET = datetime.datetime(2000, 1, 1, 0, 0)
else:
    OFFSET = datetime.datetime(1970, 1, 1, 0, 0)

# offsets of other boards are not tested and missing

class FileNotExistsError(OSError):
    pass


class StatsResult:
    """
    Stats result of a file or directory.
    """
    def __init__(self, file):
        self.mode, *_, self.size, self.atime, self.mtime, self.ctime = os.stat(file)            
        self.file = file

    @property
    def is_directory(self) -> bool:
        return bool(self.mode >> 14 & 1)

    @property
    def is_file(self) -> bool:
        return bool(self.mode >> 15 & 1)

    @property
    def dt_atime(self):
        return OFFSET + datetime.timedelta(seconds=self.atime)
    
    @property
    def dt_mtime(self):
        return OFFSET + datetime.timedelta(seconds=self.mtime)
    
    @property
    def dt_ctime(self):
        return OFFSET + datetime.timedelta(seconds=self.ctime)


def stats(file: str) -> StatsResult:
    """
    Stats result of a file or directory.
    """
    try:
        return StatsResult(file)
        
    except OSError as error:
        if error.args[0] == 2:
            raise FileNotExistsError(f"{file} does not exist.")
        else:
            raise

```

**Reply by scruss** [reply]
note: most users won't have `mip` unless they're running the newest of the new. `upip` should work

**Comment by fishjoe** [comment]
OMG this is as  good as text book, not only a detailed explanation, but also a great example of code - readability.
thank you so much!!
